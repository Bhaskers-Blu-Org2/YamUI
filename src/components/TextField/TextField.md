### Examples

One Line:
```js
<TextField
  label="First Name"
  value="Textfield filled"
  description="Small hint goes here"
/>
```

```js
<TextField
  value="Disabled"
  disabled
/>
```

```js
<TextField
  placeHolder="Textfield Placeholder..."
/>
```

```js
<TextField
  value="Error"
  errorMessage="Error hint goes here"
/>
```

```js
<TextField
  value="TODO: Icon"
/>
```

```js
<TextField
  value="TODO: Spinner"
/>
```

```js
<TextField
  placeHolder="your-website.com"
  prefix="http://"
/>
```

```js
<TextField
  placeHolder="your e-mail"
  suffix="@contoso.com"
/>
```

Multiple Line:
```js
<TextField
  value="This multiline has fixed height so scrollbars may appear."
  rows={3}
/>
```

```js
<TextField
  value="This multiline auto grows."
  autoAdjustHeight={true}
  rows={1}
/>
```

Underlined Variation
```js
<TextField
  prefix="hello"
  placeHolder="Textfield Placeholder..."
  underlined
/>
```

```js
<TextField
  errorMessage="Error hint goes here"
  placeHolder="Textfield Error"
  underlined
/>
```

Required:
```js
<TextField
  required
/>
```

onChange:
```js
<TextField
  onChange={action('change from `onChange` example')}
/>
```

onChange with debounce:
```js
<TextField
  onChangeDebounceTime={700}
  onChange={action('change from `onChange with debounce` example')}
/>
```