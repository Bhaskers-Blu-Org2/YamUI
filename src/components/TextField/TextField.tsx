/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { TextField as FabricTextField } from 'office-ui-fabric-react/lib/TextField';
import { Async } from 'office-ui-fabric-react/lib/Utilities';
import './TextField.css';

export interface TextFieldProps extends BaseComponentProps {
  /**
   * Current value of the textfield.
  */
  value?: string;

  /**
   * Placeholder text. Displayed until a value is entered.
  **/
  placeHolder?: string;

  /**
   * Error message string.
   */
  errorMessage?: string;

  /**
   * Disabled state of the textfield.
  */
  disabled?: boolean;

  /**
   * Label for the textfield.
  */
  label?: string;

  /**
   * The textfield input description
   */
  description?: string;

  /**
   * String for prefix.
  */
  prefix?: string;

  /**
   * String for suffix.
  */
  suffix?: string;

  /**
   * Use to indicate that a value must be provided to allow the containing form to be submitted.
  */
  required?: boolean;

  /**
   * Height of the field, in number of rows.
  **/
  rows?: number;

  /**
   * Whether or not to auto adjust textField height.
  **/
  autoAdjustHeight?: boolean;

  /**
   * Whether or not the textfield is underlined.
  **/
  underlined?: boolean;

  /**
   * Callback for the onChanged event.
   */
  onChange?: (newValue: any) => void;

  /**
   * Textfield will trigger onChange after users stop typing for `onChangeDebounceTime`
   * milliseconds.
   */
  onChangeDebounceTime?: number;
}

/**
 * The TextField component enables a user to type text into an app. It's typically used to capture
 * a single line of text, but can be configured to capture multiple lines of text. The text
 * displays on the screen in a simple, uniform format.
 */
export default class TextField extends React.PureComponent<TextFieldProps, {}> {
  private async: Async;

  render() {
    return (
      <FabricTextField
        value={this.props.value}
        className={this.getClasses()}
        label={this.props.label}
        description={this.props.description}
        prefix={this.props.prefix}
        suffix={this.props.suffix}
        disabled={this.props.disabled}
        required={this.props.required}
        errorMessage={this.props.errorMessage}
        placeholder={this.props.placeHolder}
        multiline={!!this.props.autoAdjustHeight || !!this.props.rows && (this.props.rows > 1)}
        rows={this.props.rows}
        resizable={false}
        autoAdjustHeight={this.props.autoAdjustHeight}
        underlined={this.props.underlined}
        onChanged={this.getOnChange()}
      />
    );
  }

  private getOnChange() {
    this.async = this.async || new Async(this);
    if (this.props.onChangeDebounceTime && this.props.onChange) {
      return this.async.debounce(this.props.onChange, this.props.onChangeDebounceTime);
    }
    return this.props.onChange;
  }

  private getClasses() {
    const { className, prefix, suffix } = this.props;
    const classes = ['y-text-field'];

    if (className) {
      classes.push(className);
    }

    if (prefix) {
      classes.push('y-text-field--with-prefix');
    }

    if (suffix) {
      classes.push('y-text-field--with-suffix');
    }

    return classes.join(' ');
  }
}
