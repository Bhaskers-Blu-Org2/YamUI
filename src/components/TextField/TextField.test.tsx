/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import TextField, { TextFieldProps } from '.';

describe('<TextField />', () => {
  let component: ShallowWrapper<TextFieldProps, {}>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<TextField />);
    });

    it('has its correct base class', () => {
      expect(component.hasClass('y-text-field')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with all options', () => {
    beforeEach(() => {
      component = shallow(
        <TextField
          value="VALUE"
          placeHolder="PLACEHOLDER"
          disabled={false}
          label="LABEL"
          description="DESCRIPTION"
          prefix="PREFIX"
          suffix="SUFFIX"
          required={true}
          rows={2}
          onChange={jest.fn()}
          onChangeDebounceTime={500}
          autoAdjustHeight={false}
        />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<TextField className="TEST_CLASSNAME" />);
    });

    it('includes that className', () => {
      expect(component.hasClass('TEST_CLASSNAME')).toBe(true);
    });

    it('still includes its base className', () => {
      expect(component.hasClass('y-text-field')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with prefix', () => {
    beforeEach(() => {
      component = shallow(<TextField prefix="PREFIX" />);
    });

    it('includes prefix className', () => {
      expect(component.hasClass('y-text-field--with-prefix')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with suffix', () => {
    beforeEach(() => {
      component = shallow(<TextField suffix="SUFFIX" />);
    });

    it('includes suffix className', () => {
      expect(component.hasClass('y-text-field--with-suffix')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with prefix and suffix', () => {
    beforeEach(() => {
      component = shallow(<TextField prefix="PREFIX" suffix="SUFFIX" />);
    });

    it('includes prefix className', () => {
      expect(component.hasClass('y-text-field--with-prefix')).toBe(true);
    });

    it('includes suffix className', () => {
      expect(component.hasClass('y-text-field--with-suffix')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
