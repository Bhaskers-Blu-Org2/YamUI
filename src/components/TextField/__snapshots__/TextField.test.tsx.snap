// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> with additional className matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={false}
  borderless={false}
  className="y-text-field TEST_CLASSNAME"
  deferredValidationTime={200}
  errorMessage=""
  multiline={false}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  resizable={false}
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
/>
`;

exports[`<TextField /> with all options matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={false}
  borderless={false}
  className="y-text-field y-text-field--with-prefix y-text-field--with-suffix"
  deferredValidationTime={200}
  description="DESCRIPTION"
  disabled={false}
  errorMessage=""
  label="LABEL"
  multiline={true}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  placeholder="PLACEHOLDER"
  prefix="PREFIX"
  required={true}
  resizable={false}
  rows={2}
  suffix="SUFFIX"
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
  value="VALUE"
/>
`;

exports[`<TextField /> with default options matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={false}
  borderless={false}
  className="y-text-field"
  deferredValidationTime={200}
  errorMessage=""
  multiline={false}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  resizable={false}
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
/>
`;

exports[`<TextField /> with prefix and suffix matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={false}
  borderless={false}
  className="y-text-field y-text-field--with-prefix y-text-field--with-suffix"
  deferredValidationTime={200}
  errorMessage=""
  multiline={false}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  prefix="PREFIX"
  resizable={false}
  suffix="SUFFIX"
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
/>
`;

exports[`<TextField /> with prefix matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={false}
  borderless={false}
  className="y-text-field y-text-field--with-prefix"
  deferredValidationTime={200}
  errorMessage=""
  multiline={false}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  prefix="PREFIX"
  resizable={false}
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
/>
`;

exports[`<TextField /> with suffix matches its snapshot 1`] = `
<TextField
  autoAdjustHeight={false}
  borderless={false}
  className="y-text-field y-text-field--with-suffix"
  deferredValidationTime={200}
  errorMessage=""
  multiline={false}
  onBeforeChange={[Function]}
  onChanged={[Function]}
  onGetErrorMessage={[Function]}
  onNotifyValidationResult={[Function]}
  resizable={false}
  suffix="SUFFIX"
  underlined={false}
  validateOnFocusIn={false}
  validateOnFocusOut={false}
  validateOnLoad={true}
/>
`;
